### PR TITLE
338 manage ws members

### PR DESCRIPTION
**Pull request by reinhardt**
UI for managing workspace members and their roles.
- add multiple members
- remove members
- assign/change roles
- remove roles
- display roles
- hide functions for unprivileged users

Compare with prototype.

### CODE REVIEW

**Comment by gyst** [comment]
Lots of security-sensitive feature work, but no test coverage at all. Please add some robot test coverage that:
- validates the extra roles visually
- but also validates the extra permissions actually, for example setting a workspace to 'consume', then elevating a single user, then validate that this user has gained 'add' permission

See for example the work I did in
https://github.com/ploneintranet/ploneintranet/commit/347a99d7b4bc750ce38f9de866377d12bf98e81f

**Comment by reinhardt** [comment]
I'm not sure the second part is within the scope of this. I just connected UI elements to existing role management features. The latter may or may not be tested - I didn't touch them.

That said, robot tests that verify the UI are still to do, as mentioned in #338.

**Comment by gyst** [comment]
@reinhardt You don't have to do heavy checks. Just something that not tautologically tests your own changes only. If you're changing permissions, other elements in the UI should change. Anchoring the test on one of those changes is sufficient.

**Comment by gyst** [comment]
@reinhardt please also merge in master so I can auto-merge on github once this is complete.

**Comment by reinhardt** [comment]
Added two robot tests and rebased/merged master.

**Comment by reinhardt** [comment]
...and two more.

File: src/ploneintranet/workspace/browser/tiles/sidebar.py [hunk 1]
```diff
@@ -127,22 +136,60 @@ def users(self):
...
+                        _(u'Member(s) removed'),
+                        self.request,
+                        'success',
+                    )
+                elif batch_function == 'role':
+                    role = self.request.get('role')
+                    default_role = ws.participant_policy.title()
+                    groups = [default_role]
+                    if role:
+                        groups.append(role)
```
Comment (gyst) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: This is in line with our original technical design, but it conflicts with the UI design. This adds custom roles on top of the default workspace roles. The UI clearly indicates that you're reassigning the user from the default role to another role.

In this implementation, if you're in a 'publish' workspace you can move users to 'consume' visually but it has no security effect whatsoever, because the user is in both ('publish', 'consume') groups and publish is a superset of consume.

The reason that this problem doesn't show up in the tests, is because they are run in a consume workspace and elevate permissions.

To summarize: with this implementation you can only elevate permissions, not assign lesser permissions than the default. IMO that should be changed so that a custom role _replaces_ rather than _adds to_ the default group.

@mattss do you agree?

File: src/ploneintranet/suite/tests/lib/keywords.robot [hunk 2]
```diff
@@ -392,6 +439,11 @@ I cannot see the document
     Click link  Documents
     Page should not contain  ${title}
 
+Allan has the option to create a document
+    I am logged in as the user allan_neece
+    I go to the Shareholder Information Workspace
```
Comment (gyst) [inline src/ploneintranet/suite/tests/lib/keywords.robot:468]: This is a 'consume' workspace which is the lowest privilege level. That's OK for testing privilege enhancement, but not for reducing access rights. For that you'd need a different workspace with a higher default policy. See other comment.

**Comment by gyst** [comment]
@reinhardt This is great work, thanks for the care you put in finishing and testing. I'm sorry it is your bad luck to expose a design inconsistency that I should have caught earlier.

File: src/ploneintranet/workspace/browser/tiles/sidebar.py (hunk 1 above)
Comment (reinhardt) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: > The UI clearly indicates that you're reassigning the user from the default role to another role.

Interesting - to me it doesn't. The fact that some users have this orange bubble with a role while regular users don't have a bubble to me conveys the idea that these users have something extra on top of what the other users have. If all other users had a bubble with the default role, maybe in a different colour, that would in my eyes be closer to what you describe.

**Comment by reinhardt** [comment]
Don't worry about it, always happy to stir things up! ;-) And thanks for the appreciation.

File: src/ploneintranet/workspace/browser/tiles/sidebar.py (hunk 1 above)
Comment (gyst) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: I get your point, visually. However the wording of the interactions makes it clear this is about assigning roles, not assigning _extra_ roles. Probably fixable by replacing the default group with the new one, instead of setting both.

Comment (reinhardt) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: > However the wording of the interactions makes it clear this is about assigning roles, not assigning extra roles. 

True.

> Probably fixable by replacing the default group with the new one, instead of setting both.

Will do (assuming @matss doesn't veto).

Comment (mattss) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: I _think_ I just broke the ability for policy exceptions in a recent commit (https://github.com/ploneintranet/ploneintranet/commit/06ac676d57546a66e99d389475f3ba2da61dc6c8) (sorry!)

Previously, this only moved uses from the currently policy to the new policy, but I changed it to move _all_ members to the new policy, because we were having persistence issues. I now think this move was foolish, so we may need to switch back to the logic where it only moves users who were in the old policy group (see diff for what it was doing originally).

If you want me to try and fix that, please let me know.

Comment (gyst) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: I think the new code is much more robust, and, as I note, it's also in line with how Cornelis has structured the UI. Simplicity is our mantra. Please keep the current behavior and let's do "replace" instead of "add" group roles.

Comment (gyst) [inline src/ploneintranet/workspace/browser/tiles/sidebar.py]: Oops hold. I now see our testcontent does set multiple roles. Lemme think.

**Comment by gyst** [comment]
So:
- make roles replacing not additive
- fix the collective.workspace permission problem
- also show 'Team Manager' as a role

**Comment by gyst** [comment]
@reinhardt Matt will take over now.

**Comment by reinhardt** [comment]
Roger that.

**Comment by gyst** [comment]
Great work all!

**Comment by gyst** [comment]
RRgh the merge broke the test. Will fix that. Later.

**Comment by pilz** [comment]
I am checking the tests now

**Comment by gyst** [comment]
@pilz I fixed this!

**Comment by pilz** [comment]
lol :) Yes, I found no errors in the tests ;) Super!
